theano/gof/cc.py

```diff
@@ -1445,22 +1464,116 @@ def compile_cmodule(self, location=None):
+            # yet computer, but we need to add the include to compute the hash.
+            filename_h = os.path.join(location, mod.hash_placeholder + '.h')
+            mod.add_include(filename_h)
         src_code = mod.code()
+        if self.c_callable:
+            filename_h = os.path.join(location, '%s.h' % mod.code_hash)
+            mod.gen_header(filename_h)
         get_lock()
         try:
             _logger.debug("LOCATION %s", str(location))
-            module = c_compiler.compile_str(
-                module_name=mod.code_hash,
-                src_code=src_code,
-                location=location,
-                include_dirs=self.header_dirs(),
-                lib_dirs=self.lib_dirs(),
-                libs=libs,
-                preargs=preargs)
-        except Exception as e:
-            e.args += (str(self.fgraph),)
-            raise
+            try:
+                module = c_compiler.compile_str(
+                    module_name=mod.code_hash,
+                    src_code=mod.code(),
+                    location=location,
+                    include_dirs=self.header_dirs(),
+                    lib_dirs=self.lib_dirs(),
+                    libs=libs,
+                    preargs=preargs)
+
+                if self.c_callable:
+                    # The main of the executable need the hash of the
+                    # shared lib.
+                    main = re.sub(mod.hash_placeholder, mod.code_hash,
+                              self.c_main())
+
+                    mod_exec = cmodule.DynamicModule()
+                    for header in self.headers():
+                        mod_exec.add_include(header)
+                    mod_exec.add_include(filename_h)
+                    mod_exec.add_support_code(main)
+                    makefile = c_compiler.make_makefile(
+                        module_name=mod.code_hash,
+                        location=location,
+                        include_dirs=self.header_dirs(),
+                        lib_dirs=self.lib_dirs(),
+                        libs=libs,
+                        preargs=preargs)
+                    f = open(os.path.join(location, 'makefile'), 'w')
+                    print(makefile, file=f)
+                    f.close()
+                    # Put the command line in the header code so that
+                    # other people know how to recompile the shared lib
+                    mod_exec.add_header_code(
+                        "//command line used to compile the shared lib: \n" +
+                        "//" + ' '.join(
+                            c_compiler.compile_command(
+                                module_name=mod.code_hash,
+                                location=location,
+                                include_dirs=self.header_dirs(),
+                                lib_dirs=self.lib_dirs(),
+                                libs=libs,
+                                preargs=preargs)[2]))
```

Comment:
I was wrong. 
I think we should keep it. Knowing the command and the flags used to compile the library can be useful.